File: php/src/Workload.php
<?php

declare(strict_types=1);

final readonly class HeavyWorkloadSummary
{
    public function __construct(
        public int $batches,
        public int $affectedRows,
        public float $elapsedSeconds,
    ) {
    }
}

final class HeavyProcess
{
    private int $completedBatches = 0;
    private ?HeavyWorkloadSummary $summary = null;
    private ?string $workerError = null;

    /** @param resource $socket */
    public function __construct(
        private readonly int $pid,
        private $socket,
    ) {
    }

    public function awaitReady(int $timeoutSeconds): void
    {
        stream_set_blocking($this->socket, true);
        stream_set_timeout($this->socket, $timeoutSeconds);

        while (true) {
            $message = $this->readMessage();
            if ($message === null) {
                $metadata = stream_get_meta_data($this->socket);
                $reason = $metadata['timed_out'] ? 'timed out' : 'closed';
                throw new RuntimeException("Heavy worker signal channel {$reason} before READY");
            }

            $this->handleMessage($message);
            if (($message['type'] ?? null) === 'ready') {
                stream_set_blocking($this->socket, false);
                return;
            }

            if ($this->workerError !== null) {
                throw new RuntimeException("Heavy worker failed: {$this->workerError}");
            }
        }
    }

    public function refreshProgress(): int
    {
        stream_set_blocking($this->socket, false);
        while (($message = $this->readMessage()) !== null) {
            $this->handleMessage($message);
        }

        if ($this->workerError !== null) {
            throw new RuntimeException("Heavy worker failed: {$this->workerError}");
        }

        return $this->completedBatches;
    }

    public function wait(): HeavyWorkloadSummary
    {
        $waitedPid = pcntl_waitpid($this->pid, $status);
        if ($waitedPid !== $this->pid) {
            throw new RuntimeException('Failed to wait for heavy worker');
        }

        $this->refreshProgress();
        fclose($this->socket);

        if (!pcntl_wifexited($status) || pcntl_wexitstatus($status) !== 0) {
            $detail = $this->workerError ?? 'unknown worker error';
            throw new RuntimeException("Heavy worker exited unsuccessfully: {$detail}");
        }

        if ($this->summary === null) {
            throw new RuntimeException('Heavy worker exited without a summary');
        }

        return $this->summary;
    }

    /** @return array<string, mixed>|null */
    private function readMessage(): ?array
    {
        $line = fgets($this->socket);
        if ($line === false) {
            return null;
        }

        $message = json_decode($line, true, flags: JSON_THROW_ON_ERROR);
        if (!is_array($message)) {
            throw new RuntimeException('Invalid heavy worker message');
        }

        return $message;
    }

    /** @param array<string, mixed> $message */
    private function handleMessage(array $message): void
    {
        $type = $message['type'] ?? null;
        if ($type === 'batch' || $type === 'ready') {
            $this->completedBatches = max(
                $this->completedBatches,
                (int) ($message['batch'] ?? 0),
            );
        } elseif ($type === 'done') {
            $this->summary = new HeavyWorkloadSummary(
                batches: (int) $message['batches'],
                affectedRows: (int) $message['affected_rows'],
                elapsedSeconds: (float) $message['elapsed_seconds'],
            );
        } elseif ($type === 'error') {
            $this->workerError = (string) ($message['message'] ?? 'unknown error');
        }
    }
}

final class Workload
{
    public static function forkHeavyWorker(Config $config): HeavyProcess
    {
        $sockets = stream_socket_pair(
            STREAM_PF_UNIX,
            STREAM_SOCK_STREAM,
            STREAM_IPPROTO_IP,
        );
        if ($sockets === false) {
            throw new RuntimeException('Failed to create heavy worker signal channel');
        }

        [$parentSocket, $childSocket] = $sockets;
        $pid = pcntl_fork();
        if ($pid === -1) {
            fclose($parentSocket);
            fclose($childSocket);
            throw new RuntimeException('Failed to fork heavy worker');
        }

        if ($pid === 0) {
            fclose($parentSocket);
            $exitCode = self::runHeavyWorker($config, $childSocket);
            fclose($childSocket);
            exit($exitCode);
        }

        fclose($childSocket);
        return new HeavyProcess($pid, $parentSocket);
    }

    public static function updateMarker(PDO $source): int
    {
        $source->exec(<<<'SQL'
            UPDATE lag_demo.marker
            SET value = value + 1, updated_at = CURRENT_TIMESTAMP(6)
            WHERE id = 1
            SQL);

        return Database::markerValue($source);
    }

    /** @param resource $socket */
    private static function runHeavyWorker(Config $config, $socket): int
    {
        $readySent = false;

        try {
            $source = Database::connect($config);
            printf(
                "Starting %d heavy UPDATE batches on Source...\n",
                $config->totalHeavyBatches(),
            );
            $workloadStarted = hrtime(true);
            $firstId = 1;
            $batchNumber = 0;
            $affectedRows = 0;
            $statement = $source->prepare(<<<'SQL'
                UPDATE lag_demo.items
                SET
                    version = version + 1,
                    payload = CONCAT(
                        SHA2(CONCAT(payload, ':', id, ':', version), 256),
                        REPEAT('z', 448)
                    )
                WHERE id BETWEEN ? AND ?
                SQL);

            while ($firstId <= $config->demoRows) {
                $lastId = min($firstId + $config->heavyBatchSize - 1, $config->demoRows);
                $batchStarted = hrtime(true);
                $statement->execute([$firstId, $lastId]);
                ++$batchNumber;
                $affectedRows += $statement->rowCount();

                printf(
                    "  heavy batch %d/%d: rows %d..=%d, affected=%d, %.2fs\n",
                    $batchNumber,
                    $config->totalHeavyBatches(),
                    $firstId,
                    $lastId,
                    $statement->rowCount(),
                    self::elapsedSeconds($batchStarted),
                );
                self::send($socket, ['type' => 'batch', 'batch' => $batchNumber]);

                if ($batchNumber === $config->markerAfterBatches) {
                    self::send($socket, ['type' => 'ready', 'batch' => $batchNumber]);
                    $readySent = true;
                }

                $firstId = $lastId + 1;
            }

            self::send($socket, [
                'type' => 'done',
                'batches' => $batchNumber,
                'affected_rows' => $affectedRows,
                'elapsed_seconds' => self::elapsedSeconds($workloadStarted),
            ]);
            return 0;
        } catch (Throwable $throwable) {
            self::send($socket, [
                'type' => 'error',
                'before_ready' => !$readySent,
                'message' => $throwable->getMessage(),
            ]);
            fprintf(STDERR, "Heavy worker error: %s\n", $throwable->getMessage());
            return 1;
        }
    }

    /**
     * @param resource $socket
     * @param array<string, mixed> $message
     */
    private static function send($socket, array $message): void
    {
        $encoded = json_encode($message, JSON_THROW_ON_ERROR) . "\n";
        if (fwrite($socket, $encoded) === false) {
            throw new RuntimeException('Failed to send heavy worker progress');
        }
        fflush($socket);
    }

    private static function elapsedSeconds(int $started): float
    {
        return (hrtime(true) - $started) / 1_000_000_000;
    }
}

File: src/workload.rs
use std::{
    sync::{
        Arc,
        atomic::{AtomicU64, Ordering},
    },
    time::Instant,
};

use anyhow::{Context, Result};
use sqlx::MySqlPool;
use tokio::{sync::oneshot, task::JoinHandle};

use crate::config::Config;

#[derive(Debug)]
pub struct HeavyWorkloadSummary {
    pub batches: u64,
    pub affected_rows: u64,
    pub elapsed_seconds: f64,
}

pub struct HeavyWorkload {
    pub ready: oneshot::Receiver<()>,
    pub progress: Arc<AtomicU64>,
    pub handle: JoinHandle<Result<HeavyWorkloadSummary>>,
}

pub fn spawn_heavy_workload(source: MySqlPool, config: Config) -> HeavyWorkload {
    let (ready_tx, ready_rx) = oneshot::channel();
    let progress = Arc::new(AtomicU64::new(0));
    let task_progress = Arc::clone(&progress);

    let handle =
        tokio::spawn(
            async move { run_heavy_workload(source, config, ready_tx, task_progress).await },
        );

    HeavyWorkload {
        ready: ready_rx,
        progress,
        handle,
    }
}

async fn run_heavy_workload(
    source: MySqlPool,
    config: Config,
    ready_tx: oneshot::Sender<()>,
    progress: Arc<AtomicU64>,
) -> Result<HeavyWorkloadSummary> {
    println!(
        "Starting {} heavy UPDATE batches on Source...",
        config.total_heavy_batches()
    );
    let workload_started = Instant::now();
    let mut first_id = 1;
    let mut batch_number = 0;
    let mut affected_rows = 0;
    let mut ready_tx = Some(ready_tx);

    while first_id <= config.demo_rows {
        let last_id = (first_id + config.heavy_batch_size - 1).min(config.demo_rows);
        let batch_started = Instant::now();
        let result = sqlx::query(
            r#"
            UPDATE lag_demo.items
            SET
                version = version + 1,
                payload = CONCAT(
                    SHA2(CONCAT(payload, ':', id, ':', version), 256),
                    REPEAT('z', 448)
                )
            WHERE id BETWEEN ? AND ?
            "#,
        )
        .bind(first_id)
        .bind(last_id)
        .execute(&source)
        .await
        .with_context(|| format!("heavy batch failed for rows {first_id}..={last_id}"))?;

        batch_number += 1;
        affected_rows += result.rows_affected();
        progress.store(batch_number, Ordering::Relaxed);

        println!(
            "  heavy batch {batch_number}/{}: rows {first_id}..={last_id}, affected={}, {:.2}s",
            config.total_heavy_batches(),
            result.rows_affected(),
            batch_started.elapsed().as_secs_f64()
        );

        if batch_number == config.marker_after_batches
            && let Some(sender) = ready_tx.take()
        {
            let _ = sender.send(());
        }

        first_id = last_id + 1;
    }

    Ok(HeavyWorkloadSummary {
        batches: batch_number,
        affected_rows,
        elapsed_seconds: workload_started.elapsed().as_secs_f64(),
    })
}

pub async fn update_marker(source: &MySqlPool) -> Result<u64> {
    sqlx::query(
        "UPDATE lag_demo.marker SET value = value + 1, updated_at = CURRENT_TIMESTAMP(6) WHERE id = 1",
    )
    .execute(source)
    .await
    .context("failed to update marker on Source")?;

    crate::database::marker_value(source).await
}
